### ISSUE TITLE
xrootd-scitokens / xrootd-multiuser 5.6.9 tends to return Permission Denied only

### ISSUE THREAD

**Issue by gsi-sfleisch** [xrootd/xrootd#2254]
We currently evaluate a setup with `xrootd-multiuser` and `xrootd-scitokens` to allow users HTTP access to their files.
`xrootd` must perform the file access as the Unix user name specified in the `mapfile` which maps the `sub` from the scitoken to the Unix user name on the machine running xrootd.

We made a simple [Vagrantfile](https://git.gsi.de/dc/XRootD-Reps/vagrant-xrootd-multiuser/-/blob/master/Vagrantfile?ref_type=heads) for this purpose and it was working with `xrootd` `5.6.8`. It no longer works with `5.6.9`.

Steps to reproduce:

This particular example requires `vagrant` and the plugin `vagrant-reload`.

```
$ vagrant up
...
$ vagrant ssh
[vagrant@rocky8 ~]$ TOKEN=$(/vagrant/provisioning/client/create_token.py)
[vagrant@rocky8 ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user1/testfile1.txt
this file belongs to user1
[vagrant@rocky8 ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user2/testfile2.txt
Unable to open /home/vagrant/xrootddata/user2/testfile2.txt; permission denied
```
Works as expected.

Updating to `xrootd` `5.6.9`

```
[vagrant@rocky8 ~]$ sudo dnf update -y
...
[vagrant@rocky8 ~]$ sudo systemctl restart xrootd-privileged@http
[vagrant@rocky8 ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user1/testfile1.txt
Unable to open /home/vagrant/xrootddata/user1/testfile1.txt; permission denied
```
It no longer works.

```
[vagrant@rocky8 ~]$ tail -f /var/log/xrootd/http/xrootd.log 
Copr. 2012 CERN IT, an HTTP implementation for the XRootD framework.
++++++ HTTP protocol initialization started.
=====> http.header2cgi Authorization authz
=====> continue /etc/xrootd/config.d/
Config continuing with file /etc/xrootd/config.d/60-osg-multiuser.cfg ...
Config warning: HTTPS functionality was not configured.
240430 15:16:31 8407 sysConfig: XRDROLE:  server
240430 15:16:31 8407 sysConfig: Configured as HTTP(s) data server.
------ HTTP protocol initialization completed.
------ xrootd http@rocky8.localdomain:80 initialization completed.
240430 15:17:06 8415 XrootdBridge: unknown.1:27@localhost login as nobody
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs
240430 15:17:06 8415 scitokens_Access: Trying token-based access control
240430 15:17:06 8415 scitokens_Access: Token not found in recent cache; parsing.
240430 15:17:06 8415 scitokens_Access: New valid token mapped_username=tokenuser1, subject=tokenuser1, issuer=https://demo.scitokens.org, authorizations=/:read,dir,stat
240430 15:17:06 8415 scitokens_Access: Request username user1
240430 15:17:06 8415 ofs_stat: unknown.1:27@localhost Unable to locate /home/vagrant/xrootddata/user1/testfile1.txt; permission denied
240430 15:17:06 8415 scitokens_Access: Trying token-based access control
240430 15:17:06 8415 scitokens_Access: Cached token mapped_username=tokenuser1, subject=tokenuser1, issuer=https://demo.scitokens.org, authorizations=/:read,dir,stat
240430 15:17:06 8415 scitokens_Access: Request username user1
240430 15:17:06 8415 ofs_open: unknown.1:27@localhost Unable to open /home/vagrant/xrootddata/user1/testfile1.txt; permission denied
240430 15:17:06 8415 XrootdXeq: unknown.1:27@localhost disc 0:00:00 (send failure)
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs
```

Please advise if it is a bug or we're merely holding it wrong.

**Comment by abh3** [comment]
Hi,
OSG supports the multi-user plugin and problems related to it should be directed to them via their problem ticketing system.
________________________________
From: gsi-sfleisch ***@***.***>
Sent: Tuesday, April 30, 2024 8:21 AM
To: xrootd/xrootd ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [xrootd/xrootd] xrootd-scitokens / xrootd-multiuser 5.6.9 tends to return Permission Denied only (Issue #2254)


We currently evaluate a setup with xrootd-multiuser and xrootd-scitokens to allow users HTTP access to their files.
xrootd must perform the file access as the Unix user name specified in the mapfile which maps the sub from the scitoken to the Unix user name on the machine running xrootd.

We made a simple Vagrantfile<https://git.gsi.de/dc/XRootD-Reps/vagrant-xrootd-multiuser/-/blob/master/Vagrantfile?ref_type=heads> for this purpose and it was working with xrootd 5.6.8. It no longer works with 5.6.9.

Steps to reproduce:

This particular example requires vagrant and the plugin vagrant-reload.

$ vagrant up
...
$ vagrant ssh
***@***.*** ~]$ TOKEN=$(/vagrant/provisioning/client/create_token.py)
***@***.*** ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user1/testfile1.txt
this file belongs to user1
***@***.*** ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user2/testfile2.txt
Unable to open /home/vagrant/xrootddata/user2/testfile2.txt; permission denied


Works as expected.

Updating to xrootd 5.6.9

***@***.*** ~]$ sudo dnf update -y
...
***@***.*** ~]$ sudo systemctl restart ***@***.***
***@***.*** ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user1/testfile1.txt
Unable to open /home/vagrant/xrootddata/user1/testfile1.txt; permission denied


It no longer works.

***@***.*** ~]$ tail -f /var/log/xrootd/http/xrootd.log
Copr. 2012 CERN IT, an HTTP implementation for the XRootD framework.
++++++ HTTP protocol initialization started.
=====> http.header2cgi Authorization authz
=====> continue /etc/xrootd/config.d/
Config continuing with file /etc/xrootd/config.d/60-osg-multiuser.cfg ...
Config warning: HTTPS functionality was not configured.
240430 15:16:31 8407 sysConfig: XRDROLE:  server
240430 15:16:31 8407 sysConfig: Configured as HTTP(s) data server.
------ HTTP protocol initialization completed.
------ xrootd ***@***.***:80 initialization completed.
240430 15:17:06 8415 XrootdBridge: ***@***.*** login as nobody
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs
240430 15:17:06 8415 scitokens_Access: Trying token-based access control
240430 15:17:06 8415 scitokens_Access: Token not found in recent cache; parsing.
240430 15:17:06 8415 scitokens_Access: New valid token mapped_username=tokenuser1, subject=tokenuser1, issuer=https://demo.scitokens.org, authorizations=/:read,dir,stat
240430 15:17:06 8415 scitokens_Access: Request username user1
240430 15:17:06 8415 ofs_stat: ***@***.*** Unable to locate /home/vagrant/xrootddata/user1/testfile1.txt; permission denied
240430 15:17:06 8415 scitokens_Access: Trying token-based access control
240430 15:17:06 8415 scitokens_Access: Cached token mapped_username=tokenuser1, subject=tokenuser1, issuer=https://demo.scitokens.org, authorizations=/:read,dir,stat
240430 15:17:06 8415 scitokens_Access: Request username user1
240430 15:17:06 8415 ofs_open: ***@***.*** Unable to open /home/vagrant/xrootddata/user1/testfile1.txt; permission denied
240430 15:17:06 8415 XrootdXeq: ***@***.*** disc 0:00:00 (send failure)
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs
240430 15:17:06 8415 multiuser_UserSentry: Anonymous client; no user set, cannot change FS UIDs


Please advise if it is a bug or we're merely holding it wrong.

**Comment by amadio** [comment]
If you have a `base_path = /` in your configuration, there was a bug introduced which is fixed in the devel branch. If you could try with the current `devel` branch and it works, we will release XRootD 5.7.0 soon. I manually tested the scitokens plugin on my machine with both demo scitokens and IAM tokens, and everything should be ok there.

**Comment by gsi-sfleisch** [comment]
Thanks, it looks like the fix you mentioned did the trick.

```
[vagrant@rocky8 ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user1/testfile1.txt
this file belongs to user1
[vagrant@rocky8 ~]$ curl -H "$TOKEN" http://127.0.0.1//home/vagrant/xrootddata/user2/testfile2.txt
Unable to open /home/vagrant/xrootddata/user2/testfile2.txt; permission denied
[vagrant@rocky8 ~]$ xrootd -v
xrootd: Symbol `_ZN7XrdInet5netIFE' has different size in shared object, consider re-linking
v5.6.9-80-g112815604
```

**Comment by amadio** [comment]
@gsi-sfleisch Great! I will close this ticket then. This issue was supposed to be fixed (see #2132), but was not completely fixed in 5.6.9. Thanks for reporting it.
